File: src/android/es.dsie.cordova.plugins.eritialabs.adserver/view/InterstitialView.java
package es.dsie.cordova.plugins.eritialabs.adserver.view;

import java.util.Random;

import android.content.Context;
import android.util.Log;
import android.view.KeyEvent;
import android.webkit.JavascriptInterface;
import android.webkit.WebView;

public class InterstitialView extends WebView {
	private static final String LOGTAG = "AdServerPlugin";
	
	private final String _TPL_URL = "%s/www/delivery/al.php?refresh=%d" +
									"&zoneid=%s" + 
									"&source=%s" + 
									"&target=%s" + 
									"&cb=%d" + 
									"&ct0=%s" + 
									"&adstext=Ads" + 
									"&closebutton=special" + 
									"&layerstyle=simple" + 
									"&align=center" + 
									"&valign=middle" + 
									"&padding=10" + 
									"&buttonsize=14" + 
									"&deferclose=5" +
									"&noborder=t";
	private final String _TPL_URL_S = "%s/www/delivery/afr.php?refresh=%d&zoneid=%s&source=%s&target=%s&cb=%d&ct0=%s";	
	
	/*
	 * ox_u += "zoneid=" + (QueryString.zoneid?QueryString.zoneid:"6");
   		ox_u += "&source=" + (QueryString.source?QueryString.source:"BacoApp");
   		ox_u += "&target=_system";
   		ox_u += "&ct0={clickurl_enc}";
   		ox_u += "&layerstyle=" + (QueryString.layerstyle?QueryString.layerstyle:"simple");
   		ox_u += "&align=" + (QueryString.align?QueryString.align:"center");
   		ox_u += "&valign=" + (QueryString.valign?QueryString.valign:"middle");
   		ox_u += "&padding=" + (QueryString.padding?QueryString.padding:"10");
   		ox_u += "&closetime=" + (QueryString.closetime?QueryString.closetime:"15");
   		ox_u += "&shifth=" + (QueryString.shifth?QueryString.shifth:"0");
   		ox_u += "&closebutton=" + (QueryString.closebutton?QueryString.closebutton:"special");
   		ox_u += "&nobg=" + (QueryString.nobg?QueryString.nobg:"t");
   		ox_u += "&noborder=" + (QueryString.noborder?QueryString.noborder:"t");
   		ox_u += "&buttonsize=" + (QueryString.buttonsize?QueryString.buttonsize:"14");
   		ox_u += "&shiftv=" + (QueryString.shiftv?QueryString.shiftv:"0");
   		ox_u += "&adstext=" + (QueryString.adstext?QueryString.adstext:"Ads");
   		ox_u += "&adstextstyle=" + (QueryString.adstextstyle?QueryString.adstextstyle:"");
   		ox_u += "&deferclose=" + (QueryString.deferclose?QueryString.deferclose:"5");
	 */
	
	
	private final String _RETURN_URL = "{clickurl_enc}";
	private final String _SOURCE = "BacoApp";
	private final String _TARGET = "_system";

	public InterstitialView(Context context) {
		super(context);
		this.getSettings().setJavaScriptEnabled(true);
		this.getSettings().setGeolocationEnabled(true);
		this.addJavascriptInterface(this, "iview");
		
		String script = 
				"javascript:window.injectPostMessage = function(e) { \n" +  
				"   if(typeof e == 'string') { \n" +  
				"		iview.doPostMessage(e); \n" +  
				"	} else { \n" +  
				"		iview.doPostMessage(e.message); \n" +  
				"	}\n" +  
				"};\n";
		
		Log.i(LOGTAG, "InterstitialView.loadAd.script=" + script);
		this.loadUrl(script);
	}
	
	public InterstitialView loadAd(String source, String domain,String zoneId,int changeInterval) {
		Random rand = new Random();
		Log.v(LOGTAG, "InterstitialView.loadAd.source:" + source);
		Log.v(LOGTAG, "InterstitialView.loadAd.domain:" + domain);
		Log.v(LOGTAG, "InterstitialView.loadAd.zoneId:" + zoneId);
		Log.v(LOGTAG, "InterstitialView.loadAd.changeInterval:" + changeInterval);

		String url = String.format(_TPL_URL,domain,changeInterval,zoneId,source,_TARGET,rand.nextInt(),_RETURN_URL);
		Log.i(LOGTAG, "InterstitialView.loadAd.URL:" + url);
		String script = 
				"javascript:setTimeout( \n" +  
				"   function(){ \n" +  
				"   	window.location.origin = 'http://adserver.eritialabs.es';  \n" +  
				"   	document.write(\"<scr\" + \"ipt type='text/javascript' src='" + url + "'></scr\" + \"ipt>\"); \n" +  
				"   },50 \n" +  
				"); \n";
		Log.i(LOGTAG, "InterstitialView.loadAd.script=" + script);
		this.loadUrl(script);
		
		/* Sin código, directamente delde el SERVER.
		String url = String.format(_TPL_URL_S,domain,changeInterval,zoneId,source,_TARGET,rand.nextInt(),_RETURN_URL);
		this.loadUrl(url);
		*/
		
		
		return this;
	}
	
	public void doClose() {
	}
	
	@JavascriptInterface	    
    public void doPostMessage(String ev) {
		Log.v(LOGTAG, "InterstitialView.doPostMessage.event=" + ev);
		if(ev instanceof String) {
			String text = (String) ev;
			Log.v(LOGTAG, "InterstitialView.doPostMessage.text=" + text);
			if(text.toLowerCase().indexOf("closed") >= 0) {
				Log.v(LOGTAG, "InterstitialView.doPostMessage.doClose()");
				this.doClose();
			}
		}
	}
	

}

File: src/ios/InterstitialView.m
#import "InterstitialView.h"
@implementation InterstitialView

@synthesize interstitialView;
@synthesize bannerView;

#define _TPL_URL = "%s/www/delivery/al.php?refresh=%d&zoneid=%s&source=%s&target=%s&cb=%d&ct0=%s&adstext=Ads&closebutton=special&layerstyle=simple&align=center&valign=middle&padding=10&buttonsize=14&deferclose=5&noborder=t";
#define _TARGET = "_system";
#define _RETURN_URL = "URL";
#define _URL_SCHEMA = "bacoApp://";

- (UIWebView*) loadAd:(UIView *) parentView:(UIWebView *) mBannerView
{
    bannerView = mBannerView;
    
    NSLog([NSString stringWithFormat:@"%f",parentView.frame.size.height]);
    NSLog([NSString stringWithFormat:@"%f",parentView.frame.size.width]);
    
    int y = 20;
    int h = parentView.frame.size.height - y + bannerView.frame.size.height;
    
    //[bannerView setHidden:TRUE];
    interstitialView = [[UIWebView alloc] initWithFrame:CGRectMake(0,y,parentView.frame.size.width,h)];
    interstitialView.delegate = self;
    interstitialView.scalesPageToFit = NO;
    
    NSURL* reqUrl = [NSURL URLWithString:@""];//bacoSchemaApp://showInterstitial];
    NSURLRequest* request = [NSURLRequest requestWithURL:reqUrl];
    [interstitialView loadRequest:request];
    
    return interstitialView;
}

-(void)webViewDidFinishLoad:(UIWebView *)webView {
    NSString *auxUrl = [NSString stringWithFormat:@"%@/www/delivery/al.php?refresh=%@&zoneid=%@&source=%@&target=%@&cb=%d&ct0=%@&adstext=Ads&closebutton=special&layerstyle=simple&align=center&valign=middle&padding=10&buttonsize=14&deferclose=5&noborder=t",domain,changeInterval,zoneId,source,@"_system",rand(),@"URL"];
    NSLog(auxUrl);
    NSString *url = [NSString stringWithFormat:@"javascript:setTimeout( \n"
                     "   function(){ \n"
                     "       window.location.origin = 'http://adserver.eritialabs.es';  \n"
                     "       document.write(\"<scr\" + \"ipt type='text/javascript' src='"
                     "%@"
                     "'></scr\" + \"ipt>\"); \n"
                     "   },50 \n"
                     "); \n",auxUrl];
    
    [interstitialView stringByEvaluatingJavaScriptFromString:url];
    
    NSString *closeScript = @"javascript:window.injectPostMessage = function(e) { \n"
    @"   if(typeof e == 'string') { \n"
    @"       window.location = 'bacoApp://doPostMessage/' + e; \n"
    @"   } else { \n"
    @"       window.location = 'bacoApp://doPostMessage/' + e.message; \n"
    @"   }\n"
    @"};\n";
    
    [interstitialView stringByEvaluatingJavaScriptFromString:closeScript];
}

-(BOOL) webView:(UIWebView *)inWeb shouldStartLoadWithRequest:(NSURLRequest *)inRequest navigationType:(UIWebViewNavigationType)inType {
    if ( inType == UIWebViewNavigationTypeLinkClicked ) {
        [[UIApplication sharedApplication] openURL:[inRequest URL]];
        return NO;
    }
    NSURL *url = [inRequest URL];
    NSString *urlStr = url.absoluteString;
    NSString *protocolPrefix = @"bacoapp://";
    
    //process only our custom protocol
    if ([[urlStr lowercaseString] hasPrefix:protocolPrefix]) {
        NSLog([NSString stringWithFormat:@"URLSchmea: %@",urlStr]);
        
        //strip protocol from the URL. We will get input to call a native method
        urlStr = [urlStr substringFromIndex:protocolPrefix.length];
        
        //Decode the url string
        urlStr = [urlStr stringByReplacingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
        
        NSArray *components = [urlStr componentsSeparatedByString:@"/"];
        NSString *function = [components objectAtIndex:0];
        
        if([function isEqualToString:@"doPostMessage"]) {
            return [self doPostMessage:components];
        }
        
        return NO;
    }
    return YES;
}

-(BOOL) doPostMessage:(NSArray *)arguments{
    NSString *message = [arguments objectAtIndex:1];
    if([[message lowercaseString] rangeOfString:@"closed"].location != NSNotFound) {
        NSLog(@"InterstitialView.doPostMessage.closed");
        [self.interstitialView setHidden:true];
        [bannerView setHidden:false];
        
    }
}

@end